### PR TITLE
Add Reveal attachment generator

### PR DESCRIPTION
**Pull request by NickEntin**
This introduces a new AardvarkReveal framework that contains utilities for generating a bug report attachment containing a [Reveal](https://revealapp.com/) bundle.

### CODE REVIEW

File: Sources/AardvarkReveal/RevealServiceBrowser.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    override init() {
+        super.init()
+
+        serviceBrowser.delegate = self
+    }
+
+    // MARK: - Internal Properties
+
+    var localService: NetService? {
```
Comment (NickEntin) [inline Sources/AardvarkReveal/RevealServiceBrowser.swift:31]: This tries to find a Reveal server running on the same device. It doesn't differentiate between multiple Reveal servers running on the same device, which may cause unexpected behavior when running apps in multiple simulators. That said, I don't think filing a bug report from a simulator is a common use case currently, so this probably isn't something we need to worry about immediately.

File: Sources/AardvarkReveal/ARKCompressionUtility.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+//
```
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKCompressionUtility.h:1]: I'm tempted to put this utility into the `Aardvark` module, since it isn't specific to the Reveal attachment generation. I could see it being useful for other attachments as well (even being able to compress the log stores, to reduce the data size of bug reports).

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+//  limitations under the License.
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Utility for building a tar archive of a Reveal bundle.
+///
+/// This is not a fully functional tar implementation and is not appropriate for use in general purpose applications.
```
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: I'd really rather this not be part of the public API, but [from what I can tell that's not easily done](https://medium.com/5-minute-break-while-coding/creating-swift-framework-with-private-objective-c-members-the-good-the-bad-and-the-ugly-4d726386644b). Hopefully the class name and headerdoc comment make it clear that this is _not_ appropriate for use as a general purpose archival tool.

One day it may make sense to build this out further and move it into the `Aardvark` module, since it could be useful for other types of attachments that have a similar bundle structure, but there's a lot of limitations of it right now that I'd rather wait until we have a real use case to deal with.

**Comment by dfed** [comment]
Tag me when this is ready for review (it's still a draft)

**Comment by NickEntin** [comment]
@dfed Still looking into a couple things, but I think this is in an okay place to review for landing as-is and iterate on it in follow up PRs. Will comment on the places I'm thinking about improving.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    failed |= write_ascii_octal(checksum, (char *)&header.checksum, 6);
+
+    if (failed) {
+        *error = [NSError errorWithDomain:ARKRevealArchiveBuilderErrorDomain
+                                     code:ARKRevealArchiveBuilderErrorCodeHeaderWriteFailure
+                                 userInfo:nil];
+    }
+
+    return header;
+}
```
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:262]: I'd feel more confident reviewing this code if this code had unit tests.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    if (v < 0) {
+        return 1;
+    }
+
+    // Start at the least significant digit and work backwards. Fill the entire length of the string, including leading
+    // zeros.
+    p += s;
+    while (s-- > 0) {
+        *--p = (char)('0' + (v & 7));
+        v >>= 3;
```
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: what are these magic numbers? the 7 and 3? Can we make them named constants?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+///         <key>localhost</key>
+///         <dict>
+///             <key>NSExceptionAllowsInsecureHTTPLoads</key>
+///             <true/>
+///             <key>NSIncludesSubdomains</key>
+///             <true/>
+///         </dict>
+///     </dict>
+/// </dict>
+/// ```
```
Comment (NickEntin) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: I wonder if we can get this to communicate with the Reveal server via the `NetService` directly, rather than HTTP requests, to avoid requiring insecure HTTP requests to `localhost`.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                                 userInfo:nil];
+    }
+
+    return header;
+}
+
+/// Writes the value `v` as an ASCII representation of an octal number of length `s` to `p`.
+///
+/// Returns `0` if successful, otherwise `1`.
+static int write_ascii_octal(int64_t v, char *p, int s) {
```
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:267]: Can these argument names be made to be human-readable (longer than 1 character)?

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+enum Class: Equatable {
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:19]: `///` comment?

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    enum CodingKeys: String, CodingKey {
+        case name = "n"
+        case superclass = "s"
+    }
+
+}
+
+// MARK: -
+
+struct ApplicationState {
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:60]: `///` comment?

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+NSString *const ARKRevealArchiveBuilderErrorDomain = @"ARKRevealArchiveBuilderErrorDomain";
+NSInteger const ARKRevealArchiveBuilderErrorCodeAlreadyComplete = 1;
+NSInteger const ARKRevealArchiveBuilderErrorCodeDataAllocationFailure = 2;
+NSInteger const ARKRevealArchiveBuilderErrorCodeHeaderWriteFailure = 3;
+
+
+@interface ARKRevealArchiveBuilder ()
+
+@property (nonatomic) NSMutableData *archiveData;
```
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:53]: I want to follow up here with having this write the archive data to disk, rather than storing it in memory. As is, this could potentially cause some OOM problems during the attachment generation. The complexity of that comes on the compression side, where it needs to page in data from disk while it's compressing it.

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+extension Screens: Decodable {}
+
+// MARK: -
+
+struct Object: Equatable {
+
+    var identifier: Int
+
+    var `class`: Class
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:86]: 😬 can we call this `className`? If not, why not?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+///                   `───────────────────────────'
+/// ```
+///
+/// It is important to avoid changing anything while the app state is being captured (during the `Capturing App State`
+/// phase). Once the app state is captured, the generator will begin capturing snapshots of each drawable object (views,
+/// layers, etc.) on the screen. Once the generator has captured a snapshot of the main screen (marked by a call to the
+/// delegate's `revealAttachmentGeneratorDidCaptureMainScreenSnapshot` method), it is okay to display a waiting
+/// indicator **as long as all of the existing views/layers remain in the hierarchy** (for example, you can display a
+/// new window on top of the main window or present a modal view controller, depending on the use case). Once the
+/// generator has completed bundling the Reveal file, user interaction can be restored.
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:130]: one hell of a comment. well done.

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 13]
```diff
@@ -0,0 +1,518 @@
...
+/// ```
+@objc(ARKRevealAttachmentGenerator)
+public final class RevealAttachmentGenerator: NSObject {
+
+    // MARK: - Life Cycle
+
+    public override init() {
+        super.init()
+
+        serviceBrowser.startSearching()
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:161]: 😬 why are we kicking off work from `init`?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 14]
```diff
@@ -0,0 +1,518 @@
...
+
+    // MARK: - Public Methods
+
+    /// Begins the process of capturing the current app state and bundling the results into an attachment containing a
+    /// Reveal file.
+    ///
+    /// - parameter completionQueue: The queue on which to call the `completion`.
+    /// - parameter completion: The completion handler to call when the capture is complete. Called with `nil` when the
+    /// capture could not successfully be completed.
+    @available(iOS 11, *)
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:206]: I thought our minimum version was iOS 12?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 15]
```diff
@@ -0,0 +1,518 @@
...
+    ) {
+        guard let service = serviceBrowser.localService else {
+            completionQueue.async {
+                completion(nil)
+            }
+            return
+        }
+
+        let port = service.port
+        guard port != -1 else {
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:220]: can we turn `-1` into a named constant?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 16]
```diff
@@ -0,0 +1,518 @@
...
+            }
+            task.resume()
+        }
+
+        dispatchGroup.notify(queue: completionQueue) {
+            completion()
+        }
+    }
+
+    private func applicationName() -> String {
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:457]: could this method be made static?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 17]
```diff
@@ -0,0 +1,518 @@
...
+        baseRevealURL: URL,
+        urlSession: URLSession,
+        archiveBuilder: ARKRevealArchiveBuilder,
+        completionQueue: DispatchQueue,
+        completion: @escaping (_ success: Bool) -> Void
+    ) throws {
+        // Create the bundle based on the application name, clamped to a reasonable length so that the file paths don't
+        // go over 100 characters (a limitation of the archive builder).
+        let appName = applicationName()
+        let bundleName = "\(appName.prefix(50)).reveal"
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: 50 is significantly lower than 100 😅

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift [hunk 18]
```diff
@@ -0,0 +1,518 @@
...
+        )
+        try (propertiesPlist as NSDictionary).write(to: propertiesFileURL)
+        try archiveBuilder.addFile(
+            atPath: "\(bundleName)/Properties.plist",
+            with: try Data(contentsOf: propertiesFileURL)
+        )
+        try? FileManager.default.removeItem(at: propertiesFileURL)
+    }
+
+}
```
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:481]: I'm struggling with how this class can be modified confidently in the future. Can we add some unit or contractural tests?

**Review by dfed — COMMENTED** [review]
Solid stuff! Lots of thoughts, but we're getting somewhere 🙂

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 17 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: It is indeed. The longest file path right now is for the resource images:

```
<app_name>.reveal/Resources/2692677334741286913#1.png
```

That's the app name, followed by ".reveal/Resources/", a numeric identifier, "#", either "0" or "1", and the file extension. I'm assuming the identifiers are always 19 digits, although I'm not 100% sure of that. That gives us 43 characters plus a null terminator. So in theory this limit could be 56 characters, but 50 felt like a reasonable limit that gives a little bit of padding in case I'm wrong about the length of the identifier being capped at 19.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m (hunk 5 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: This is the conversion to octal (base 8). Doing a bitwise and with 7 gets you the last three bits (`7` in decimal is `111` in binary). Add that to `'0'` and you get that number in ASCII (`'0'` through `'7'`). Then you shift by 3 bits to get the next least significant digit.

Maybe it would be clearer to replace `7` with `0b111`?

File: AardvarkSample/AardvarkSample/AppDelegate.swift [hunk 19]
```diff
@@ -59,10 +68,98 @@ class SampleAppDelegate: UIResponder, UIApplicationDelegate {
...
+extension SampleAppDelegate: ARKEmailBugReporterPromptingDelegate {
+
+    func showBugReportingPrompt(
+        for configuration: ARKEmailBugReportConfiguration,
+        completion: @escaping ARKEmailBugReporterCustomPromptCompletionBlock
+    ) {
+        UIApplication.shared.beginIgnoringInteractionEvents()
+
+        revealAttachmentGenerator?.captureCurrentAppState(completionQueue: .main) { attachment in
+            UIApplication.shared.endIgnoringInteractionEvents()
```
[the remark this answers is not in the document]
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:112]: Correct. Or, at least, it's definitely a bug if it isn't.

File: .github/workflows/ci.yml [hunk 20]
```diff
@@ -53,3 +53,5 @@ jobs:
         run: bundle exec pod lib lint --verbose --fail-fast AardvarkLoggingUI.podspec --include-podspecs=CoreAardvark.podspec
       - name: Lint AardvarkMailUI Podspec
         run: bundle exec pod lib lint --verbose --fail-fast AardvarkMailUI.podspec --include-podspecs="{CoreAardvark,Aardvark}.podspec"
+      - name: Lint AardvarkReveal Podspec
```
[the remark this answers is not in the document]
Comment (NickEntin) [reply .github/workflows/ci.yml:56]: Great question. I think in the long term it definitely makes sense to grow out an ecosystem of components in separate repos, with this repo holding the core components and maybe a link to the other repos in the README. The main reason for having it here for now is to have it in the sample app. Maybe it's still worth starting that pattern now and moving it to a separate repo with its own sample app...

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 17 above)
Comment (dfed) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: Got it. The file path limit is def odd. How you're tackling this makes sense.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m (hunk 5 above)
Comment (dfed) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: `0b111` would be a bit more clear. I'd also strongly recommend wring tests for this code.

Comment (sethfri) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: I'm just a fly on the wall, but I think this could also be made clearer by taking the explanation you've provided here on GitHub and turning that into more granular comments (rather than one comment at the start of this code block). That way, you're walking the reader through it a bit more, and readers will have to brush up less on their bit manipulation 🙏🏻

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h (hunk 3 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: Correct, this problem would go away if this were written in Swift. I find it a lot easier to work with things that require specific memory layout (the header struct in particular here) in Objective-C than Swift though. If you have suggestions for how this could be ported to Swift without requiring a bunch of unsafe pointers, I'm definitely open to it.

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 12 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:130]: Thanks, this took quite a while to write up 😂

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 13 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:161]: Fair question. I was trying to make it hard to get wrong (i.e. if you forget to start it early). Maybe that's too magical and I should expose the start/stop methods.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h (hunk 3 above)
Comment (dfed) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: 😂 I mean... yeah it's certainly harder to write this kind of code in Swift. I don't have much in the way of suggestions here.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m [hunk 21]
```diff
@@ -0,0 +1,285 @@
...
+
+        _isComplete = YES;
+
+        return _archiveData;
+    }
+}
+
+#pragma mark - Private Methods
+
+- (struct header_posix_ustar)_headerForPath:(NSString *)path
```
[the remark this answers is not in the document]
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: Addressed in 462da40

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m (hunk 4 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:252]: Added some tests in 462da40

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m (hunk 5 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:269]: Addressed in 462da40 and aa78c12

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m (hunk 7 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: Addressed in 77aec64

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 16 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in c528e39

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 6 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: @tonyarnold Do you know if communicating via the `NetService` directly would remove the need for this requirement? Or is there an alternate option for communicating with the Reveal server? I don't have much experience with `NetService`, so not sure what options are available.

Comment (tonyarnold) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: My understanding is that `NetService` is only used for broadcasting availability of services and clients of services. You can't communicate over them directly, so as far as I'm aware, Reveal's `NetService` would be broadcasting details of the HTTP endpoint and the port it is running on to net service browsers. 

I'll have a look when I'm back at my desk later today, and verify exactly what we're doing.

Comment (tonyarnold) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: Side note: by _golly_ I wish that we offered alternatives here!

We've been a bit hamstrung in terms of some of the nice new technology Apple has added in recent releases - our support for older iOS releases means that we can't use many of them in the server component. I'll need to reassess that now that we're moving our minimum iOS and tvOS versions up to 10.x releases.

Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: Ahh, makes sense! Sending the requests over HTTP seems to work great with the exception of this requirement. It would be awesome to get a little bit cleaner alternative, but I definitely don't think this would be a blocker.

**Comment by NickEntin** [comment]
Came across the file coordinator's utility for creating a ZIP archive from a directory and switched over to using that in favor of the custom TAR builder. This should resolve a lot of the open comments right now.

File: AardvarkSample/AardvarkSample/AppDelegate.swift [hunk 22]
```diff
@@ -59,10 +68,98 @@ class SampleAppDelegate: UIResponder, UIApplicationDelegate {
...
+
+// MARK: -
+
+extension SampleAppDelegate: ARKEmailBugReporterPromptingDelegate {
+
+    func showBugReportingPrompt(
+        for configuration: ARKEmailBugReportConfiguration,
+        completion: @escaping ARKEmailBugReporterCustomPromptCompletionBlock
+    ) {
+        UIApplication.shared.beginIgnoringInteractionEvents()
```
[the remark this answers is not in the document]
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:109]: - [ ] Add comment about ignoring interaction events while bundle is being generated

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 14 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:206]: Not yet, #92 is still open.

- [ ] Remove unnecessary availability checks once deployment target bump lands

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 15 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:220]: - [ ] Turn missing port check into named constant

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 18 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:475]: - [ ] Add tests for attachment generator

File: AardvarkSample/AardvarkSample/AppDelegate.swift (hunk 22 above)
[the remark this answers is not in the document]
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:111]: Addressed in dc6ea26

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 15 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in 33d1f73

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 14 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in f9a5614

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift (hunk 18 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:490]: Added some integration tests for the generator's callbacks in different scenarios. Still want to get one more big one in to test its interaction with the archive builder.

Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:495]: Pretty happy with the test setup I've got now. Probably a few more edge cases that could be covered, but I think it's got pretty good coverage already.

**Comment by NickEntin** [comment]
@dfed I think this is ready for another review pass when you get a chance. I'm looking into moving this extension into its own repo, but I'd like to still go through the review process here and get everything approved before copying it over.

**Comment by NickEntin** [comment]
@dfed New repo is good to go! Once this is PR approved I'll copy the code over to the new repo and make it public.

File: Sources/AardvarkReveal/ARKCompressionUtility.m [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+//
+
+#import "ARKCompressionUtility.h"
+
+#import <zlib.h>
+
+
+@implementation ARKCompressionUtility
+
+// The GZIP functionality in this class is based on the GZIP framework, which has the following license:
```
[the remark this answers is not in the document]
Comment (dfed) [reply Sources/AardvarkReveal/ARKCompressionUtility.m:24]: alternatively, why not create a dependency on the GZIP library?

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+        case let .baseClass(name: name):
+            try values.encode(name, forKey: .name)
+        case let .subclass(name: name, superclass: superclass):
+            try values.encode(name, forKey: .name)
+            try values.encode(superclass, forKey: .superclass)
+        }
+    }
+
+    enum CodingKeys: String, CodingKey {
+        case name = "n"
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:63]: If this case name is required due to compatibility with another library or model, would be good to state as much in a comment

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+enum Class: Equatable {
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:19]: repeating myself, but `///` comments would be great even on internal classes

File: Sources/AardvarkReveal/ApplicationState.swift [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+            try values.encode(contents, forKey: .value)
+
+        case let .object(object):
+            try values.encode("NSObject", forKey: .type)
+            try values.encode(object, forKey: .value)
+        }
+    }
+
+    enum CodingKeys: String, CodingKey {
+        case type = "t"
```
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:179]: same thing here. if these shortened coding keys are necessary, let's explain why in a comment

File: Sources/AardvarkRevealTests/ApplicationStateTests.swift [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+//  limitations under the License.
+//
+
+import XCTest
+
+@testable import AardvarkReveal
+
+final class ApplicationStateTests: XCTestCase {
+
+    func testClass() throws {
```
Comment (dfed) [inline Sources/AardvarkRevealTests/ApplicationStateTests.swift:23]: love these test, but naming the test methods in a more elucidating way would be great. Consider writing test methods using the form `test_thingBeingTested_stateExpectation` to make long-term maintenance easier.

**Review by dfed — APPROVED** [review]
LGTM, but `///` comments on internal types would be helpful for future maintainers.

File: Sources/AardvarkReveal/ApplicationState.swift (hunk 24 above)
Comment (NickEntin) [reply Sources/AardvarkReveal/ApplicationState.swift:70]: Added a comment at the top of the file. This whole file mirrors the format used in the JSON representation of the application state included in a `.reveal` file.

**Comment by NickEntin** [comment]
I'm going to close this PR now. This source code will all be copied over to a new repo as-is, along with a new demo app and documentation. I'll post back here with a link to the new repo once it's live. Thanks for the reviews!

**Comment by dfed** [comment]
Just fat-fingered that deletion! Giant touch target on iPad 🤦‍♂️
